File: mcts-vanilla.py
import math
import random

class Node:
    def __init__(self, parent=None):
        self.visits = 0
        self.value = 0.0
        self.parent = parent
        self.children = []

    def is_fully_expanded(self):
        # Allow up to 5 children for better branching
        return len(self.children) >= 5

    def is_terminal(self):
        # Consider a node terminal after 10 visits for simplicity
        return self.visits >= 10

    def add_child(self, child):
        self.children.append(child)

# Selection - Uses Upper Confidence Bound for Trees (UCT)
def select(node):
    current = node
    while current.children:
        best_child = None
        best_value = -float('inf')
        for child in current.children:
            uct_value = (child.value / (child.visits + 1e-6) +
                         1.414 * math.sqrt(math.log(current.visits + 1) / (child.visits + 1e-6)))
            if uct_value > best_value:
                best_value = uct_value
                best_child = child
        current = best_child
    return current

# Expansion - Adds a child node
def expand(node):
    if not node.is_fully_expanded():
        new_child = Node(parent=node)
        node.add_child(new_child)

# Simulation - Simulate a random playthrough
def simulate(node):
    # Random outcome for simplicity, imagine simulating the actual game here
    return 1.0 if random.randint(0, 1) == 0 else 0.0

# Backpropagation - Update the value and visits
def backpropagate(node, reward):
    current = node
    while current is not None:
        current.visits += 1
        current.value += reward
        current = current.parent

# MCTS Function - Combines the steps
def MCTS(root, iterations):
    random.seed()
    for _ in range(iterations):
        selected_node = select(root)
        if not selected_node.is_terminal():
            expand(selected_node)
        expanded_node = selected_node if not selected_node.children else selected_node.children[-1]
        reward = simulate(expanded_node)
        backpropagate(expanded_node, reward)
    return root

if __name__ == "__main__":
    root = Node()
    MCTS(root, 1000)
    print(f"Root visits: {root.visits}, value: {root.value}")
    for child in root.children:
        print(f"Child visits: {child.visits}, value: {child.value}")

File: mcts_vanilla.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <limits>
#include <cstdlib>
#include <ctime>

// Define a simple structure for nodes
struct Node {
    int visits;
    double value;
    Node* parent;
    std::vector<Node*> children;

    Node(Node* parent = nullptr) : visits(0), value(0), parent(parent) {}

    bool isFullyExpanded() const {
        // Allow more children for better exploration
        return children.size() >= 5; // Allow up to 5 children for better branching
    }

    bool isTerminal() const {
        // Check if node is terminal by limiting depth or checking visits
        return visits >= 10; // Consider a node terminal after 10 visits for simplicity
    }

    void addChild(Node* child) {
        children.push_back(child);
    }
};

// Selection - Uses Upper Confidence Bound for Trees (UCT)
Node* select(Node* root) {
    Node* current = root;
    while (!current->children.empty()) {
        Node* bestChild = nullptr;
        double bestValue = -std::numeric_limits<double>::infinity();
        for (Node* child : current->children) {
            double uctValue = child->value / (child->visits + 1e-6) +
                             1.414 * std::sqrt(std::log(current->visits + 1) / (child->visits + 1e-6)); // Increased exploration coefficient
            if (uctValue > bestValue) {
                bestValue = uctValue;
                bestChild = child;
            }
        }
        current = bestChild;
    }
    return current;
}

// Expansion - Adds a child node
void expand(Node* node) {
    if (!node->isFullyExpanded()) {
        Node* newChild = new Node(node);
        node->addChild(newChild);
    }
}

// Simulation - Simulate a random playthrough
double simulate(Node* node) {
    // Random outcome for simplicity, imagine simulating the actual game here
    return rand() % 2 == 0 ? 1.0 : 0.0;
}

// Backpropagation - Update the value and visits
void backpropagate(Node* node, double reward) {
    Node* current = node;
    while (current != nullptr) {
        current->visits++;
        current->value += reward;
        current = current->parent;
    }
}

// MCTS Function - Combines the steps
Node* MCTS(Node* root, int iterations) {
    srand(static_cast<unsigned>(time(0)));
    for (int i = 0; i < iterations; ++i) {
        Node* selectedNode = select(root);
        if (!selectedNode->isTerminal()) {
            expand(selectedNode);
        }
        Node* expandedNode = selectedNode->children.empty() ? selectedNode : selectedNode->children.back();
        double reward = simulate(expandedNode);
        backpropagate(expandedNode, reward);
    }
    return root;
}

int main() {
    Node* root = new Node();
    MCTS(root, 1000);
    std::cout << "Root visits: " << root->visits << ", value: " << root->value << std::endl;
    for (Node* child : root->children) {
        std::cout << "Child visits: " << child->visits << ", value: " << child->value << std::endl;
    }
    return 0;
}